### PR TITLE
Skills code refactor/cleanup.

### PR DESCRIPTION
**Pull request by Ghommie**
## About The Pull Request
Mostly cleaning up the skills code: 
- removed the dozens defines set for each skill in favor of generic and easier to manage ones.
- coded in some bare skill modifiers feature for temporary skill changes. These are currently only used in place of the aSL promotion and demotion leadership skill modification on overwatch.dm.
- replaced the various `user.mind?.mind.cm_skills && user.mind.cm_skills.skill` etcetera checks with macros, due to their increased complexity with the addition of skill modifiers and standarizing them a bit more.
- a couple more macros.
- Split off job skillsets into a different datum as a way to preserve skill modifiers on some potential non-admin job changes. Though I'll add an arg to the inherit_skillset() as mean to override them to null, if you want to.
- Slight clean up on make_reference_datum() while I was in adding the skillsets global list.
- Added in an omnighost var to resolve issues regarding (human) observers not passing potential skill checks mainly found on examine(), without resorting to isghost(user) everytime. It's set true on mob.ghostize() and false on mind.transfer_to().

As a consequence, the balance of one or two pair of skillgated fumbling times may have been slighty altered. C&C welcome, pretty sure there is place for improvements.

Edit: was suggested to make skills/skillsets use cached tags, will take some time to carry on said improvement. Edit: Nope, that would cause various issues with admins doing varediting shenanigeans

## Why It's Good For The Game
Cutting out snowflake defines. Adding a basic frame for skill modifications (always good to prevent edge cases), allowing ghosts to get more detailed infos from several objects and, if you want to deem it so, code improvement.

## Changelog
:cl:
fix: Ghosts should be able to get skillgated infos on examine() from more objects.
refactor: Refactored skills code. The gameplay should remain unaffected minus a few cases.
/:cl:

### CODE REVIEW

File: code/__HELPERS/global_lists.dm [hunk 1]
```diff
@@ -40,9 +40,7 @@ GLOBAL_LIST_EMPTY(randomized_pill_icons)
 		GLOB.body_types_list[B.name] = B
 
 	// Surgery Steps - Initialize all /datum/surgery_step into a list
-	for(var/T in subtypesof(/datum/surgery_step))
-		var/datum/surgery_step/S = new T
-		GLOB.surgery_steps += S
+	GLOB.surgery_steps = init_subtypes(/datum/surgery_step)
```
Comment (DominikPanic) [inline code/__HELPERS/global_lists.dm]: Unrelated changes

File: code/__HELPERS/global_lists.dm [hunk 2]
```diff
@@ -158,18 +154,36 @@ GLOBAL_LIST_EMPTY(randomized_pill_icons)
...
 	if(!istype(L))
 		L = list()
-	for(var/path in subtypesof(prototype))
+	for(var/path in (subtypes ? subtypesof(prototype) : typesof(prototype)) - blacklist)
 		L += new path()
 	return L
+
+//returns a list of paths to every subtype of prototype (excluding prototype and blacklisted)
+//if no list/L is provided, one is created.
+/proc/init_paths(prototype, list/L, blacklist, subtypes = TRUE)
```
Comment (DominikPanic) [inline code/__HELPERS/global_lists.dm]: Make this a different PR

File: code/datums/jobs/job/imperium.dm [hunk 3]
```diff
@@ -147,7 +147,7 @@
 	H.equip_to_slot_or_del(new /obj/item/cell/lasgun/M43, SLOT_IN_BACKPACK)
 	H.equip_to_slot_or_del(new /obj/item/cell/lasgun/M43, SLOT_IN_BACKPACK)
 	H.equip_to_slot_or_del(new /obj/item/defibrillator, SLOT_IN_BACKPACK)
-
```
Comment (DominikPanic) [inline code/datums/jobs/job/imperium.dm]: Unrelated whitespace changes

File: code/modules/vehicles/multitile/hardpoints.dm [hunk 4]
```diff
@@ -40,7 +40,7 @@ Currently only has the tank hardpoints
 	. = ..()
 	var/status = obj_integrity <= 0.1 ? "broken" : "functional"
 	var/span_class = obj_integrity <= 0.1 ? "<span class = 'danger'>" : "<span class = 'notice'>"
-	if((user?.mind?.cm_skills && user.mind.cm_skills.engineer >= SKILL_ENGINEER_METAL) || isobserver(user))
+	if(!isxeno(user) && HAS_SKILL_LEVEL(user, SKILL_ENGINEERING, SKILL_LEVEL_NOVICE))
```
Comment (DominikPanic) [inline code/modules/vehicles/multitile/hardpoints.dm]: Xenos should never have skills

File: code/__DEFINES/skills.dm [hunk 5]
```diff
@@ -171,3 +40,85 @@
...
+#define DEFAULT_SKILLSET /datum/skillset/pfc
+
+//D(irect) as in they don't perform the mind and skills datums existence sanity check. Use these only if their existence has been already checked.
+#define D_GET_SKILL_MOD(user, skill) (user.mind.cm_skills.skill_mods ? user.mind.cm_skills.skill_mods[skill] : SKILL_LEVEL_NONE)
+#define D_GET_SKILL(user, skill) (user.mind.cm_skills.omnighost ? SKILL_LEVEL_MASTER : CLAMP((user.mind.cm_skills.skillset ? user.mind.cm_skills.skillset[skill] : SKILL_LEVEL_NONE) + D_GET_SKILL_MOD(user, skill), SKILL_LEVEL_MIN, SKILL_LEVEL_MAX))
+#define D_HAS_SKILL_LEVEL(user, skill, level) (D_GET_SKILL(user, skill) >= level)
+#define D_GET_SKILL_MOD_ID(user, skill, id) (user.mind.cm_skills.skill_mod_ids && user.mind.cm_skills.skill_mod_ids[skill] ? user.mind.cm_skills.skill_mod_ids[skill[id]] : null)
+#define D_GET_SKILL_DIFF(user, skill, level) abs(level - D_GET_SKILL(user, skill))
+
+#define D_ADD_SKILL_MOD(user, skill, id, mod) \
```
Comment (psykzz) [inline code/__DEFINES/skills.dm]: There doesn't seem to be any gain to putting these in as a DEFINE. Can this not just be a proc on /mob or even /living (whatever should have skills).

Comment (Ghommie) [reply code/__DEFINES/skills.dm]: Are you saying all these macros should be turned into procs or only the multi-line ones?

File: code/datums/jobs/job/imperium.dm (hunk 3 above)
Comment (Ghommie) [reply code/datums/jobs/job/imperium.dm]: I did attempt to reset, diff and stage it to no avail just right now.
Forgive me for being a slowpoke using shabby git GUIs on old potatoes in the attempt to save up time and effort. This is what I get for not doing --ignore whitespace on git bash.

Comment (DominikPanic) [reply code/datums/jobs/job/imperium.dm]: Are you using DM to code? Other editors dont' do this.

Comment (Ghommie) [reply code/datums/jobs/job/imperium.dm]: Yes, I'm temporarily using a different potato and never bothered to switch to a better program (tried visualcode on the other one, but it was too memory-expensiveto work with), even though I have notepad++ installed and could use it.

Comment (DominikPanic) [reply code/datums/jobs/job/imperium.dm]: Sublime is all you need. It's better and more efficient than DM.

File: code/__DEFINES/skills.dm (hunk 5 above)
Comment (Rohesie) [reply code/__DEFINES/skills.dm]: @psykzz , procs are good if you need to redefine them, but unlike in other sane languages BYOND has an issue with proc overhead. Meaning we a) have a global limit on how many active procs there can be and b) they actually decrease performance.
Having defines/macros for stuff that needs to redefining is good and should be encouraged as long as it doesn't obliterate readability.

File: code/datums/jobs/job/imperium.dm (hunk 3 above)
Comment (Rohesie) [reply code/datums/jobs/job/imperium.dm]: Notepad++ is good as well. Poke me on discord if you need help solving these.

**Comment by DominikPanic** [comment]
Travis

**Comment by DominikPanic** [comment]
Was this tested locally?

**Comment by Ghommie** [comment]
Yes, this is pretty much tested locally now @DominikPanic.

File: code/game/objects/machinery/computer/skills.dm [hunk 6]
```diff
@@ -21,6 +21,12 @@
 	var/sortBy = "name"
 	var/order = 1 // -1 = Descending - 1 = Ascending
 
+/proc/I_runtime_on_rails()
+	var/list/L = list()
+	var/chumpo = "fetusin"
+	var/chumpo2 = "fetus"
+	L[chumpo[chumpo2]] = "???"
+	return L
```
Comment (DominikPanic) [inline code/game/objects/machinery/computer/skills.dm]: What the actual fuck is this

File: code/datums/skills.dm [hunk 7]
```diff
@@ -1,529 +1,658 @@
...
+	var/list/skill_list = ALL_SKILL_TYPES
+
+	var/list/skill_mods //contains the total skill modifier values. refer to __DEFINES/skills.dm for its usage in the macros.
+	var/list/skill_mod_ids //contains the applied skill modifier IDs, sorted by skill. used in macros too.
+
+	//used on /mob/proc/check_skills()
+	var/cached_skills_dat
+
+	//used to allow observers to bypass possible skill checks mainly found on examine().
+	var/omnighost = FALSE
```
Comment (DominikPanic) [inline code/datums/skills.dm]: Perhaps a better name like "bypass"? Assuming this will allow admins to remove all restrictions easily

File: code/game/objects/machinery/computer/skills.dm (hunk 6 above)
Comment (Ghommie) [reply code/game/objects/machinery/computer/skills.dm]: Some garbage lines I almost subconsciously wrote down as I tried to figure out an issue I later resolved but then forgot to remove. kek.

File: code/datums/mind.dm [hunk 8]
```diff
@@ -74,13 +74,17 @@
 
 	SSnano.user_transferred(current, new_character) // transfer active NanoUI instances to new user
 
+	if(cm_skills)
+		cm_skills.omnighost = FALSE
+		current.verbs -= /mob/proc/check_skills
```
[the remark this answers is not in the document]
Comment (Ghommie) [reply code/datums/mind.dm]: Skills datum are tied to the mind, not mobs.
This is done with the intention of potentially allowing any "skilled" mob to review their skills, while still preventing the ones without. Instead of just tying this up to humans only (ergo human ghosts should be able to use the verb too now).

**Comment by spookydonut** [comment]
@Ghommie merge conflict

**Comment by spookydonut** [comment]
Stale closing
